### DISCUSSION TITLE
Occasional Lag spikes when using fluent search

### DISCUSSION THREAD

**Discussion by rsmyst** [adirh3/Fluent-Search#247] [General]
I attend classes all morning. So, when I have something like google meet running in the background and use fluent search, the typing kinda skips over some letters. I mean that all letters appear but some letters don't pop up continuosly. It just appears suddenly with the next letter instead of appearing individually letter by letter if you get what I mean.

**Expected behavior**
No lag spikes

**Desktop (please complete the following information):**
 - Windows 10 Version: 20H1
 - Fluent Search Version: 0.9.86.0
 - Dual core CPU if that matters (i7-7500U)

**Comment by adirh3** [comment]
@rsmyst Can you check if using "Disable instant search" fixes the issue?
In Settings -> In-app search

**Comment by MakeshVineeth** [comment]
I must say even I'm facing this, but didn't report it earlier, as I thought that lag is because of my potato system 😅 Looks like disabling instant search made it bit better but not perfect.

**Comment by rsmyst** [comment]
> @rsmyst Can you check if using "Disable instant search" fixes the issue?
> In Settings -> In-app search

I attended a zoom session and tried fluent search and would like to confirm that, it does in fact still lag then. It misses up to 4-6 characters
Something else I noticed was soon after I restarted, I waited for FS to open and elevate to admin.
After that, I launched it with the shortcut but after typing around 30-40 characters in quick succession, it just crashed.

**Comment by rsmyst** [comment]
> I must say even I'm facing this, but didn't report it earlier, as I thought that lag is because of my potato system 😅 Looks like disabling instant search made it bit better but not perfect.

That didn't help me, unfortunately.. Maybe marginally...

**Comment by adirh3** [comment]
@rsmyst @MakeshVineeth can you check if disabling Settings -> Processes -> Search in app content helps?
Thanks

**Comment by rsmyst** [comment]
> @rsmyst @MakeshVineeth can you check if disabling Settings -> Processes -> Search in app content helps?
> Thanks

I have tried this and can confirm it still doesn't seem to help. On a sidenote, my fans rev up massively when having a meeting + using fs search. Like a fan speed spike.
I've done some digging around using task manager and found that 
1. My cpu runs at about 20% with a meeting running on average
2. Upon launching FS search and using, it spikes to 70-80% and then returns to around 35-40%

**Comment by adirh3** [comment]
> > @rsmyst @MakeshVineeth can you check if disabling Settings -> Processes -> Search in app content helps?
> > Thanks
> 
> I have tried this and can confirm it still doesn't seem to help. On a sidenote, my fans rev up massively when having a meeting + using fs search. Like a fan speed spike.
> I've done some digging around using task manager and found that
> 
> 1. My cpu runs at about 20% with a meeting running on average
> 2. Upon launching FS search and using, it spikes to 70-80% and then returns to around 35-40%

Upon launching Fluent Search for the first time or just opening it with default shortcut?

Also, I don't think the input delay issue is related to CPU spikes unless they're near the 100%, if Fluent Search causes CPU spikes when it's running in the background then something is wrong.

I will try checking with other users on Discord if any have similar issue and then we might find the root cause.

**Comment by rsmyst** [comment]
> Upon launching Fluent Search for the first time or just opening it with default shortcut?
> 
> Also, I don't think the input delay issue is related to CPU spikes unless they're near the 100%, if Fluent Search causes CPU spikes when it's running in the background then something is wrong.
> 
> I will try checking with other users on Discord if any have similar issue and then we might find the root cause.

Okay clarification. Opening it with shortcut and not for first time. It happens when opening it suddenly after a good amount of break since the last use of fluent search.
FS is not causing spikes while in Background. Only when using the shortcut and searching something.

**Comment by MakeshVineeth** [comment]
Disabling process search doesn't help me much with speed. But it looks like the lags are caused only when FS has to generate different set of results. When I search using a specific tag e.g. settings or chromium, it doesn't seem to lag for me. @rsmyst Can you confirm the same?

**Comment by rsmyst** [comment]
> Disabling process search doesn't help me much with speed. But it looks like the lags are caused only when FS has to generate different set of results. When I search using a specific tag e.g. settings or chromium, it doesn't seem to lag for me. @rsmyst Can you confirm the same?

Yea I think I noticed something similar as well. I had previously chalked it up to random stutters but maybe it's generating new results that causes it. A massive cpu spike for 2 seconds

**Comment by adirh3** [comment]
@rsmyst Let me know if you notice any improvements in 0.9.87.0

**Comment by MakeshVineeth** [comment]
I'm still facing that text lag when I'm typing too fast i.e., each letter will appear with a delay as mentioned in OP. Using FS version 0.9.88.1

**Comment by adirh3** [comment]
> I'm still facing that text lag when I'm typing too fast i.e., each letter will appear with a delay as mentioned in OP. Using FS version 0.9.88.1

Can you check if this is related to a specific setting? 
Try disabling each search app and see if it changes anything.
Also, do you have adaptive color on?

**Comment by MakeshVineeth** [comment]
> > I'm still facing that text lag when I'm typing too fast i.e., each letter will appear with a delay as mentioned in OP. Using FS version 0.9.88.1
> 
> Can you check if this is related to a specific setting?
> Try disabling each search app and see if it changes anything.
> Also, do you have adaptive color on?

It seems like Files search is the one that's causing the text lag. After I disable it, and try to type fast, I no longer face that lag. Also, I don't have adaptive color on.

**Comment by rsmyst** [comment]
> @rsmyst Let me know if you notice any improvements in 0.9.87.0

It doesn't seem too much smoother to me personally even in this version

**Comment by rsmyst** [comment]
> > > I'm still facing that text lag when I'm typing too fast i.e., each letter will appear with a delay as mentioned in OP. Using FS version 0.9.88.1
> > 
> > 
> > Can you check if this is related to a specific setting?
> > Try disabling each search app and see if it changes anything.
> > Also, do you have adaptive color on?
> 
> It seems like Files search is the one that's causing the text lag. After I disable it, and try to type fast, I no longer face that lag. Also, I don't have adaptive color on.

too bad we can't disable file search since that's the point of the app... 
Good luck for fixing it! Take your time!

**Comment by adirh3** [comment]
@rsmyst @mv just wondering you guys have SSD or HDD?

**Comment by MakeshVineeth** [comment]
250GB Samsung SSD

**Comment by rsmyst** [comment]
> @rsmyst @mv just wondering you guys have SSD or HDD?

512 gb SATA SSD

**Comment by MakeshVineeth** [comment]
@adirh3 Just now I faced something weird and thought I'd like to share. Looks like FS has caused massive CPU hikes. So much that rest of the system has become slower temporarily, even the typing has become slower in rest of the applications. This has happened to me twice today: Once when launching the FS for the first time, and once in the background. I did not check the Task Manager but I could tell because when I faced that lag, I immediately closed both **Discord App** and **FS**. Boom! All of a sudden, system and text typing speed came back to normal. I could hardly think if it is caused by Discord as I've been running that app since quite a long time 🤔

**Comment by adirh3** [comment]
> @adirh3 Just now I faced something weird and thought I'd like to share. Looks like FS has caused massive CPU hikes. So much that rest of the system has become slower temporarily, even the typing has become slower in rest of the applications. This has happened to me twice today: Once when launching the FS for the first time, and once in the background. I did not check the Task Manager but I could tell because when I faced that lag, I immediately closed both **Discord App** and **FS**. Boom! All of a sudden, system and text typing speed came back to normal. I could hardly think if it is caused by Discord as I've been running that app since quite a long time

It's due to In-app search try disabling it.
On startup it's slow and it's a known issue, it will improve over time.

**Comment by MakeshVineeth** [comment]
**EDIT:** Here are my observations after testing for a while.
- Fluent Search with Fluent search File Indexer, text typing lags.
- Enabled Everything Instead, minor lags at first but okay for the most part.
- With Windows Search enabled, no visible lags at all so far.

Facing the text typing lag even in 0.9.88.4; But I observed something: Setting the File Indexer to Windows Search seems to smoothen up the things a little bit.

**Comment by adirh3** [comment]
@MakeshVineeth @rsmyst Can you please try the following -

- Make sure Use GPU is enabled
- Make sure Adaptive Color is disabled
- Change the max search results to a small number (5-10)
- Change the search delay to High
- Rebuild index using Fluent Search
- Adding more directories to ignore in the File indexer and then rebuild index

After testing all the above, please let me know if any of them reduced lags, it will help me know the right direction for mitigating this issue.

**Comment by MakeshVineeth** [comment]
> @MakeshVineeth @rsmyst Can you please try the following -
> 
> * Make sure Use GPU is enabled
> * Make sure Adaptive Color is disabled
> * Change the max search results to a small number (5-10)
> * Change the search delay to High
> * Rebuild index using Fluent Search
> * Adding more directories to ignore in the File indexer and then rebuild index
> 
> After testing all the above, please let me know if any of them reduced lags, it will help me know the right direction for mitigating this issue.

**EDIT:** Setting Search Delay to Auto doesn't seem to help me much with lags, it's still same as before.

Hello, I've followed all of those steps. It doesn't seem to help in reducing the lags. Even with max search results kept to 8, CPU still seem to hit 100%. Not to mention, I've even added OneDrive, Music, Videos, Program Files (along with x86) to folder exceptions.

**Comment by rsmyst** [comment]
> @MakeshVineeth @rsmyst Can you please try the following -
> 
> * Make sure Use GPU is enabled
> * Make sure Adaptive Color is disabled
> * Change the max search results to a small number (5-10)
> * Change the search delay to High
> * Rebuild index using Fluent Search
> * Adding more directories to ignore in the File indexer and then rebuild index
> 
> After testing all the above, please let me know if any of them reduced lags, it will help me know the right direction for mitigating this issue.

Update, I changed search delay to auto and the performance is much smoother over all. Changing it to anything other than that option causes massive lag spikes though. 
Keeping search delay to Auto also doesn't cause CPU spikes
Also, as makesh has said, using everything seemingly makes the experience smoother

**Comment by adirh3** [comment]
@rsmyst @MakeshVineeth Changed this to discussion, since performance will always improve but depending on specs some may experience lags, this thread will help people finding tips to optimize performance based on their PC specs.
I will keep improving the overall performance and tune default settings based on the average user experience.

**Reply by MakeshVineeth** [reply]
Nice, looking forward to upcoming performance updates :)

**Comment by MakeshVineeth** [comment]
@rsmyst Are you running on a dual-core by any chance?

**Reply by rsmyst** [reply]
> @rsmyst Are you running on a dual-core by any chance?

actually, I am. That is a factor?

**Reply by MakeshVineeth** [reply]
Maybe, I'm also using a dual core.

**Comment by rsmyst** [comment]
Performance seems to have somehow gotten worse in the latest update.
I have no clue why. There's a lot of delay here and there along with slower searches for some reason.
Good luck fixing this!

**Reply by adirh3** [reply]
If you experience delay with search, please check the `Search Delay` setting in Settings -> Search.
It might be set to Auto, you might want to set it to `Low` to have lower delay.

**Reply by rsmyst** [reply]
I've tried. But the issue seems to be actual lag and not search delay only

**Comment by zipityzi** [comment]
I was having some similar issues: I tried searching "Notepad" and then deleting the search term and typing "Firefox". Often, it would take 500ms+ for the Firefox results to appear and at the same time, the typing lagged inside the Fluent Search window.

The fix for me was as mentioned by adirh3: I had forgotten I'd set the max search results to 75 (!) after FS was searching for a file, so I was trying to re-prioritize it with some machine learning repetition. I set the maximum results to 10 results and FS is now **lightning fast** again (stable v 0.9.90.999). 

The listed performance improvements are excellent. Perhaps a small warning could temporarily appear to remind users, within the app, that "such a high setting may significantly decrease FS performance"? Or a little caution icon? This is definitely user error, though.

I noticed the "max results" tip was not in the Discord FAQ, so I searched here and lo behold, the fix was already available.

Or, the documentation / wiki, too!

**Reply by adirh3** [reply]
Thanks @ikjadoon for the info, I will update the FAQ in Discord.
Also, FYI the lags when setting high max search results is something I am trying to fix in the nightly builds.
